core/drivers/crypto/caam/acipher/caam_ecc.c

```diff
@@ -0,0 +1,724 @@
+ */
+static TEE_Result do_shared_secret(struct drvcrypt_secret_data *sdata)
+{
+	TEE_Result ret = TEE_ERROR_GENERIC;
+	enum caam_status retstatus = CAAM_FAILURE;
+	enum caam_ecc_curve curve = CAAM_ECC_UNKNOWN;
+	struct ecc_keypair *inprivkey = sdata->key_priv;
+	struct ecc_public_key *inpubkey = sdata->key_pub;
+	struct caam_ecc_keypair ecckey = { };
+	struct caam_jobctx jobctx = { };
+	uint32_t *desc = NULL;
+	uint32_t desclen = 0;
+	struct caamdmaobj secret = { };
+	uint32_t pdb_sgt_flags = 0;
+
+#ifdef CFG_CAAM_64BIT
+#define MAX_DESC_SHARED 10
+#else
+#define MAX_DESC_SHARED 7
+#endif
+
+	ECC_TRACE("ECC Shared Secret");
+
+	/* Verify first if the curve is supported */
+	curve = get_caam_curve(inpubkey->curve);
+	if (curve == CAAM_ECC_UNKNOWN)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	/* Allocate the job descriptor */
+	desc = caam_calloc_desc(MAX_DESC_SHARED);
+	if (!desc) {
+		ret = TEE_ERROR_OUT_OF_MEMORY;
+		goto out;
+	}
+
+	/* Convert the Private key to local key */
+	retstatus = do_keypair_conv(&ecckey, inprivkey, sdata->size_sec);
+	if (retstatus != CAAM_NO_ERROR) {
+		ret = caam_status_to_tee_result(retstatus);
+		goto out;
+	}
+
+	/* Convert the Public key to local key */
+	retstatus = do_keypub_conv(&ecckey, inpubkey, sdata->size_sec);
+	if (retstatus != CAAM_NO_ERROR) {
+		ret = caam_status_to_tee_result(retstatus);
+		goto out;
+	}
+
+	/*
+	 * ReAllocate the secret result buffer with a maximum size
+	 * of the secret size if not cache aligned
+	 */
+	ret = caam_dmaobj_output_sgtbuf(&secret, sdata->secret.data,
+					sdata->secret.length, sdata->size_sec);
+	if (ret)
+		goto out;
+
+	if (secret.sgtbuf.sgt_type)
+		pdb_sgt_flags |= PDB_SGT_PKDH_SECRET;
+
+	caam_dmaobj_cache_push(&secret);
+
+	/* Build the descriptor using Predifined ECC curve */
+	caam_desc_init(desc);
+	caam_desc_add_word(desc, DESC_HEADER(0));
+	caam_desc_add_word(desc, PDB_SHARED_SECRET_PD1 | PDB_ECC_ECDSEL(curve)
+				 | pdb_sgt_flags);
+	/* Public key */
+	caam_desc_add_ptr(desc, ecckey.xy.paddr);
+	/* Private key */
+	caam_desc_add_ptr(desc, ecckey.d.paddr);
+	/* Output secret */
+	caam_desc_add_ptr(desc, secret.sgtbuf.paddr);
+
+	caam_desc_add_word(desc, SHARED_SECRET(ECC));
+	desclen = caam_desc_get_len(desc);
+	caam_desc_update_hdr(desc, DESC_HEADER_IDX(desclen, desclen - 1));
+
+	ECC_DUMPDESC(desc);
+
+	jobctx.desc = desc;
+
+	retstatus = caam_jr_enqueue(&jobctx, NULL);
+
+	if (retstatus == CAAM_NO_ERROR) {
+		sdata->secret.length = caam_dmaobj_copy_to_orig(&secret);
+
+		ECC_DUMPBUF("Secret", sdata->secret.data, sdata->secret.length);
+
+		ret = caam_status_to_tee_result(retstatus);
+	} else {
+		ECC_TRACE("CAAM Status 0x%08" PRIx32, jobctx.status);
+		ret = job_status_to_tee_result(jobctx.status);
+	}
+
+out:
+	caam_free_desc(&desc);
+	do_keypair_free(&ecckey);
+	caam_dmaobj_free(&secret);
+
+	return ret;
+}
+
+/*
+ * Registration of the ECC Driver
+ */
+static struct drvcrypt_ecc driver_ecc = {
+	.alloc_keypair = &do_allocate_keypair,
```

Comment:
Using a `&` in front of the name of the function is redundant and we're trying to avoid that.

Reply:
Ok